nixos/mattermost: add secretFile option for declarative configs (again)

stuebinm opened the pull request:
This adds an option `services.mattermost.secretFile`, which may be used in conjunction with `services.mattermost.mutableConfig` set to false to avoid placing secret parts of the configuration into the world-readable nix store.

This change was previously PR https://github.com/NixOS/nixpkgs/pull/117046, but a mistaken push operation left github confused enough to close that one & mark it as merged.

Paging @pinpox for review

###### Motivation for this change
https://github.com/NixOS/nixpkgs/issues/116421



###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

pinpox commented:
I tested this in a VM and can confirm the values from `secretFile` are applied correctly.
The configuration used for the test was:
```nix
          system.activationScripts.create-mm-secrets = ''
            cat > /var/lib/mattermost/secrets.json << END

            {
              "PasswordSettings": {
                "MinimumLength": 7
              }
            }

            END
          '';

          # Test-Specific configuration for the module
          services.mattermost = {
            enable = true;
            mutableConfig = false;
            siteUrl = "http://localhost";

            secretFile = "/var/lib/mattermost/secrets.json";

            extraConfig = {
              PasswordSettings = {
                MinimumLength = 5;
                Lowercase = false;
                Number = false;
                Uppercase = false;
                Symbol = false;
              };
            };
          };

```
Which results in `MinimumLength` to be set to 7 correctly (overriding the 5 from extraConfig). 

Since all mattermost configuration options can be overridden by environment variables, wouldn't it make more sense to add a `environmentFile` like other services do? I think it would make the module simpler and get rid of merging values

pinpox commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -184,6 +184,25 @@ in
+          is set to <literal>true</literal>, then this option has no effect.
+        '';
+      };
 
```
```suggestion
      environmentFile = mkOption {
        type = types.nullOr types.path;
        default = null;
        description = ''
          Environment file (see <literal>systemd.exec(5)</literal>
          "EnvironmentFile=" section for the syntax) to define variables for
          Mattermost. Environment variables will override settings in config.json. See 
          https://docs.mattermost.com/configure/configuration-settings.html#environment-variables for reference
        '';
      };
```

pinpox replied:
Unit further down would need, but I can't suggest the change there.
```nix
        EnvironmentFile = [  ] ++ optional (cfg.environmentFile != null) cfg.environmentFile;

```

stuebinm replied:
That's indeed be a simpler solution (though tbh I quite like having the secret file having essentially the same nice hierarchical structure as the config itself, but that's just personal preference).

I'm not sure why I didn't do it that way — probably I wasn't aware mattermost could be configured via environment variables, or perhaps that even got added in the year since I originally wrote this PR?

stuebinm commented:
I've changed this to work as @pinpox suggested, and added a (very simple) nixos test to show that it works & mattermost behaves correctly in ignoring the value set via nix if it's also given in the environment file.

stuebinm commented:
… so, is anything still to be done? or could someone review/merge it?

stuebinm commented:
so this has been lying around for over a year now. I've just rebased it again (& brought it in line with the newer markdown documentation).

Is there anything blocking this being merged? If not, could someone merge it? If yes, could someone tell me what's blocking it?

K900 commented:
The change seems fine to me, but the commit history has accumulated a lot of noise over time (sorry!). Can you just squash it down to a single commit?

K900 commented:
Also, I _think_ (but don't quote me on this) the `mkIf` is unnecessary [here](https://github.com/NixOS/nixpkgs/pull/162479/files#diff-3517cc8e7468615ccea52f1b8fb62377b14733d2e0a08ae5c7d390650402b57eR340), as `EnvironmentFile = null;` will do the right thing anyway.

(Github is having a stroke, so I can't comment directly on the line)

stuebinm committed:
mattermost: add environmentFile option to allow declarative secrets

This adds an option `services.mattermost.environmentFile`, intended to be
useful especially when `services.mattermost.mutableConfig` is set to `false`.
Since all mattermost configuration options can also be set by environment
variables, this allows managing secret configuration values in a declarative
manner without placing them in the nix store.

K900 commented:
@ofborg test mattermost

stuebinm commented:
> The change seems fine to me, but the commit history has accumulated a lot of noise over time (sorry!). Can you just squash it down to a single commit?

done

> Also, I _think_ (but don't quote me on this) the `mkIf` is unnecessary [here](https://github.com/NixOS/nixpkgs/pull/162479/files#diff-3517cc8e7468615ccea52f1b8fb62377b14733d2e0a08ae5c7d390650402b57eR340), as `EnvironmentFile = null;` will do the right thing anyway.

I've just ran the tests with this, and you're right. I've removed it from the module.
